pkg/controller/master/nodedrain/nodedrain_controller.go

```diff
@@ -0,0 +1,293 @@
+	ctlcorev1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/rest"
+	v1 "kubevirt.io/api/core/v1"
+
+	"github.com/harvester/harvester/pkg/config"
+	ctlnode "github.com/harvester/harvester/pkg/controller/master/node"
+	kubevirtv1 "github.com/harvester/harvester/pkg/generated/controllers/kubevirt.io/v1"
+	longhornv1beta1 "github.com/harvester/harvester/pkg/generated/controllers/longhorn.io/v1beta1"
+	"github.com/harvester/harvester/pkg/util"
+	"github.com/harvester/harvester/pkg/util/drainhelper"
+)
+
+const (
+	nodeDrainController = "node-drain-controller"
+	defaultWorkloadType = "VirtualMachineInstance"
+)
+
+// ControllerHandler to drain nodes.
+// the controller checks if node has any replicas which may be the only working replica for a VM and attempts to shutdown the VM
+// before the drain is initiated. This ensures that no data is lost when the instance-managers are terminated as
+// part of the drain process
+type ControllerHandler struct {
+	nodes                        ctlcorev1.NodeClient
+	nodeCache                    ctlcorev1.NodeCache
+	virtualMachineInstanceCache  kubevirtv1.VirtualMachineInstanceCache
+	virtualMachineInstanceClient kubevirtv1.VirtualMachineInstanceClient
+	virtualMachineClient         kubevirtv1.VirtualMachineClient
+	virtualMachineCache          kubevirtv1.VirtualMachineCache
+	longhornVolumeCache          longhornv1beta1.VolumeCache
+	longhornReplicaCache         longhornv1beta1.ReplicaCache
+	restConfig                   *rest.Config
+	context                      context.Context
+}
+
+func Register(ctx context.Context, management *config.Management, options config.Options) error {
+	nodes := management.CoreFactory.Core().V1().Node()
+	vmi := management.VirtFactory.Kubevirt().V1().VirtualMachineInstance()
+	vmc := management.VirtFactory.Kubevirt().V1().VirtualMachine()
+	lhv := management.LonghornFactory.Longhorn().V1beta1().Volume()
+	lhr := management.LonghornFactory.Longhorn().V1beta1().Replica()
+	ndc := &ControllerHandler{
+		nodes:                        nodes,
+		nodeCache:                    nodes.Cache(),
+		virtualMachineInstanceCache:  vmi.Cache(),
+		virtualMachineInstanceClient: vmi,
+		virtualMachineClient:         vmc,
+		virtualMachineCache:          vmc.Cache(),
+		longhornReplicaCache:         lhr.Cache(),
+		longhornVolumeCache:          lhv.Cache(),
+		restConfig:                   management.RestConfig,
+		context:                      ctx,
+	}
+
+	nodes.OnChange(ctx, nodeDrainController, ndc.OnNodeChange)
+	return nil
+}
+
+// OnNodeChange handles reconcile logic for node drains
+func (ndc *ControllerHandler) OnNodeChange(key string, node *corev1.Node) (*corev1.Node, error) {
+	if node == nil || node.DeletionTimestamp != nil {
+		return node, nil
+	}
+	_, ok := node.Annotations[drainhelper.DrainAnnotation]
+	_, forced := node.Annotations[drainhelper.ForcedDrain]
```

Comment:
Do we want to check whether the value is `"true"` or `"false"` for both annotations?

Reply:
not sure, the UI is only going to send one possible value.